### PR TITLE
adding support for analog outputs for ROS2

### PR DESCRIPTION
**Pull request by AlexisFetet**
This is a cleaned #144 PR.
This PR adds support on ROS2 Foxy for analog outputs.
This PR adds a new `auto_enable` parameter to enable outputs on output detection.

### CODE REVIEW

**Comment by clalancette** [comment]
Due to the recent releases and movement, this needs a rebase.  Sorry about that.  Do you mind doing that rebase, then I'll review?  Thanks.

**Comment by mintar** [comment]
In case you're not familiar with git, here is how to do it in your case:

```bash
git reset --hard origin/rolling
git cherry-pick 0f6d963 3672b53
git push --force-with-lease
```

**Comment by AlexisFetet** [comment]
Hi,
@clalancette this should be good now.
Thanks for the help @mintar really appreciated.

File: phidgets_analog_outputs/package.xml [hunk 1]
```diff
@@ -0,0 +1,26 @@
+<?xml version="1.0"?>
+<package format="2">
+  <name>phidgets_analog_outputs</name>
+  <version>2.1.2</version>
```
Comment (clalancette) [inline phidgets_analog_outputs/package.xml]: This number needs to match the current release, which on the `rolling` branch would be 2.3.0.

File: phidgets_analog_outputs/CHANGELOG.rst [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+1.0.5 (2022-02-17)
+------------------
+
+1.0.4 (2021-10-22)
+------------------
+
+1.0.3 (2021-09-29)
+------------------
+* Add Analog Outputs (`#103 <https://github.com/ros-drivers/phidgets_drivers/issues/103>`_)
+* Contributors: Carsten Plasberg
```
Comment (clalancette) [inline phidgets_analog_outputs/CHANGELOG.rst]: Let's remove this file for now.  We'll automatically generate it when we do a new release with this package in place.

File: phidgets_msgs/srv/SetAnalogOutput.srv [hunk 3]
```diff
@@ -0,0 +1,6 @@
+# Sets the state of a digital output.
```
Comment (clalancette) [inline phidgets_msgs/srv/SetAnalogOutput.srv]:
```suggestion
# Sets the state of an analog output.
```

File: phidgets_api/include/phidgets_api/analog_output.hpp [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#ifndef PHIDGETS_API_ANALOG_OUTPUT_H
+#define PHIDGETS_API_ANALOG_OUTPUT_H
```
Comment (clalancette) [inline phidgets_api/include/phidgets_api/analog_output.hpp]:
```suggestion
#ifndef PHIDGETS_API_ANALOG_OUTPUT_HPP
#define PHIDGETS_API_ANALOG_OUTPUT_HPP
```

File: phidgets_api/include/phidgets_api/analog_output.hpp [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+
+    void setEnabledOutput(int enabled) const;
+
+  private:
+    PhidgetVoltageOutputHandle ao_handle_;
+};
+
+}  // namespace phidgets
+
+#endif  // PHIDGETS_API_ANALOG_OUTPUT_H
```
Comment (clalancette) [inline phidgets_api/include/phidgets_api/analog_output.hpp]:
```suggestion
#endif  // PHIDGETS_API_ANALOG_OUTPUT_HPP
```

File: phidgets_api/include/phidgets_api/analog_outputs.hpp [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#ifndef PHIDGETS_API_ANALOG_OUTPUTS_H
+#define PHIDGETS_API_ANALOG_OUTPUTS_H
```
Comment (clalancette) [inline phidgets_api/include/phidgets_api/analog_outputs.hpp]:
```suggestion
#ifndef PHIDGETS_API_ANALOG_OUTPUTS_HPP
#define PHIDGETS_API_ANALOG_OUTPUTS_HPP
```

File: phidgets_api/include/phidgets_api/analog_outputs.hpp [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+    void setEnabledOutput(int index, int enabled) const;
+
+  private:
+    uint32_t output_count_;
+    std::vector<std::unique_ptr<AnalogOutput>> aos_;
+};
+
+}  // namespace phidgets
+
+#endif  // PHIDGETS_API_ANALOG_INPUTS_H
```
Comment (clalancette) [inline phidgets_api/include/phidgets_api/analog_outputs.hpp]:
```suggestion
#endif  // PHIDGETS_API_ANALOG_INPUTS_HPP
```

File: phidgets_analog_outputs/src/analog_outputs_ros_i.cpp [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <memory>
```
Comment (clalancette) [inline phidgets_analog_outputs/src/analog_outputs_ros_i.cpp:31]:
```suggestion
#include <functional>
#include <memory>
#include <string>
```

**Review by clalancette — CHANGES_REQUESTED** [review]
I've left a bunch of very minor comments.  Once those are resolved, I think this is good to merge.  Thanks for the contribution!

**Comment by AlexisFetet** [comment]
Hi @clalancette,
I applied the suggested changes, please let me know if anything else need to be changed.

**Review by clalancette — APPROVED** [review]
I noticed one more header guard fix that I missed last review.  I've fixed that myself.

With that, this looks good to me!  Once the CI comes back green, I'll go ahead and merge it.  Thanks for the contribution.

**Pull request merged**